File: 1275. Find Winner on a Tic Tac Toe Game.java
// Easy
// Simulation
// Time:O(1),Space:O(1), the size of the board is fixed
// https://leetcode.cn/problems/find-winner-on-a-tic-tac-toe-game/

class Solution {
  public String tictactoe(int[][] moves) {
      int[][] board = new int[3][3];
      // simulate the moves
      for (int i = 0; i < moves.length; i++) {
          int row = moves[i][0];
          int col = moves[i][1];
          // Determine the player: A(1) or B(-1)
          board[row][col] = (i % 2 == 0) ? 1 : -1;
          // Check if the cur player has won
          if (checkWin(board, row, col)) {
              return (i % 2 == 0) ? "A" : "B";
          }
      }
      // if no winner, determine the game state
      return moves.length == 9 ? "Draw" : "Pending";
  }

  private boolean checkWin(int[][] board, int row, int col) {
      int player = board[row][col];
      // check the cur row
      if (board[row][0] + board[row][1] + board[row][2] == 3 * player) return true;
      // check the cur col
      if (board[0][col] + board[1][col] + board[2][col] == 3 * player) return true;
      // check the main diagonal
      if (row == col && board[0][0] + board[1][1] + board[2][2] == 3 * player) return true;
      // check the anti-diagonal
      if (row + col == 2 && board[0][2] + board[1][1] + board[2][0] == 3 * player) return true;
      return false;
  }
}

File: 1001-2000/1275. Find Winner on a Tic Tac Toe Game.py
# Easy
# Simulation, Array
# Time:O(n), Space:O(1)
# https://leetcode.cn/problems/find-winner-on-a-tic-tac-toe-game/

from typing import *

class Solution:
    def tictactoe(self, moves: List[List[int]]) -> str:
        board = [['' for _ in range(3)] for _ in range(3)]
        for i, (r, c) in enumerate(moves):
            board[r][c] = 'A' if i % 2 == 0 else 'B'
        
        def check_winner():
            for row in board:
                if row[0] == row[1] == row[2] != '':
                    return row[0]
            
            for c in range(3):
                if board[0][c] == board[1][c] == board[2][c] != '':
                    return board[0][c]
            
            if board[0][0] == board[1][1] == board[2][2] != '':
                return board[0][0]
            if board[0][2] == board[1][1] == board[2][0] != '':
                return board[0][2]
            
            return None
        
        winner = check_winner()
        if winner:
            return winner
        
        return "Draw" if len(moves) == 9 else "Pending"
